File: leetcode/0284-peeking-iterator.java
// Java Iterator interface reference:
// https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html

/**
 * PeekingIterator extends the Iterator interface with a peek() operation.
 * This allows viewing the next element without consuming it from the iterator.
 */
class PeekingIterator implements Iterator<Integer> {
    // The underlying iterator that provides the actual elements
    private Iterator<Integer> iterator;
  
    // Flag to track whether we have a peeked element cached
    private boolean hasPeeked;
  
    // Cache for storing the peeked element
    private Integer peekedElement;

    /**
     * Constructor initializes the PeekingIterator with a given iterator.
     * @param iterator The underlying iterator to wrap
     */
    public PeekingIterator(Iterator<Integer> iterator) {
        this.iterator = iterator;
        this.hasPeeked = false;
        this.peekedElement = null;
    }

    /**
     * Returns the next element in the iteration without advancing the iterator.
     * Multiple consecutive calls to peek() should return the same element.
     * @return The next element that would be returned by next()
     */
    public Integer peek() {
        // If we haven't peeked yet, fetch the next element and cache it
        if (!hasPeeked) {
            peekedElement = iterator.next();
            hasPeeked = true;
        }
        // Return the cached peeked element
        return peekedElement;
    }

    /**
     * Returns the next element in the iteration and advances the iterator.
     * @return The next element in the iteration
     */
    @Override
    public Integer next() {
        // If we haven't peeked, directly return from the underlying iterator
        if (!hasPeeked) {
            return iterator.next();
        }
      
        // If we have peeked, return the cached element and reset the peek state
        Integer result = peekedElement;
        hasPeeked = false;
        peekedElement = null;
        return result;
    }

    /**
     * Returns true if the iteration has more elements.
     * @return true if there are more elements, false otherwise
     */
    @Override
    public boolean hasNext() {
        // We have next if either we have a peeked element or the underlying iterator has next
        return hasPeeked || iterator.hasNext();
    }
}

File: 0284-peeking-iterator/0284-peeking-iterator.ts
/**
 * // This is the Iterator's API interface.
 * // You should not implement it, or speculate about its implementation.
 *
 * interface Iterator {
 *     next(): number;
 *     hasNext(): boolean;
 * }
 */

/**
 * @param iterator
 */
class PeekingIterator {
    private iterator: Iterator;
    private hasPeeked: boolean;
    private peekedElement: number;

    constructor(iterator: Iterator) {
        this.iterator = iterator;
        this.hasPeeked = false;
        this.peekedElement = 0;
    }

    /**
     * @return {number}
     */
    peek(): number {
        // 아직 peek하지 않았다면 다음 값을 미리 저장
        if (!this.hasPeeked) {
            this.peekedElement = this.iterator.next();
            this.hasPeeked = true;
        }

        return this.peekedElement;
    }

    /**
     * @return {number}
     */
    next(): number {
        // peek한 값이 없다면 iterator에서 바로 가져옴
        if (!this.hasPeeked) {
            return this.iterator.next();
        }

        // 저장된 peek 값을 반환
        this.hasPeeked = false;
        return this.peekedElement;
    }

    /**
     * @return {boolean}
     */
    hasNext(): boolean {
        // peek한 값이 있거나 iterator에 다음 값이 있으면 true
        return this.hasPeeked || this.iterator.hasNext();
    }
}